msprime/provenance.py

```diff
@@ -68,3 +76,131 @@ def get_environment():
+            # the array is human readable
+            return {
+                '__ndarray__': obj.tolist(),
+                'dtype': obj.dtype.str
+            }
+
+        elif isinstance(obj, numpy.number):
+            # Some numbers come through as numpy types
+            return {
+                '__npgeneric__': str(obj),
+                'dtype': obj.dtype.str,
+            }
+
+        try:
+            ret = obj.asdict()
+            cls = obj.__class__
+            ret['__class__'] = f'{cls.__module__}.{cls.__name__}'
+            return ret
+        except AttributeError:
+            raise TypeError(f'Object of type {obj.__class__.__name__} '
+                            f'is not JSON serializable, provide an `asdict` method '
+                            f'that gives the objects args')
+
+    @staticmethod
+    def decode(s):
+        def hook(obj):
+            if '__function__' in obj:
+                return types.FunctionType(
+                    marshal.loads(base64.b64decode(obj['__function__'])),
+                    {}, 'func', obj['defaults'], ())
+            elif '__ts__' in obj:
+                return tskit.TableCollection.fromdict(obj['__ts__']).tree_sequence()
+            elif '__ndarray__' in obj:
+                return numpy.asarray(obj['__ndarray__'], dtype=obj['dtype'])
+            elif '__npgeneric__' in obj:
+                return numpy.array([obj['__npgeneric__']]).astype(obj['dtype'])[0]
+            elif '__class__' in obj:
+                module, cls = obj['__class__'].rsplit('.', 1)
+                module = importlib.import_module(module)
+                del obj['__class__']
+                return getattr(module, cls)(**obj)
+            return obj
+        return json.JSONDecoder(object_hook=hook).decode(s)
+
+
+def parse_provenance(ts, index=None):
+    """
+    Convert the specified provenance to tuple of the command used and dict suitable
+    for calling the command again e.g `msprime.simulate(**parse_provenance(ts)[1])`
+    """
+    if index is None:
+        index = ts.num_provenances - 1
+    print(index)
+    prov = ts.provenance(index).record
+    ret = ProvenanceEncoderDecoder.decode(prov)
+    if ret['software']['name'] != 'msprime':
+        raise ValueError(f'Only msprime provanances can be parsed,'
+                         f' found {ret["software"]["name"]}')
+    command = ret['parameters']['command']
+    del ret['parameters']['command']
+    return command, ret['parameters']
+
+
+def _human_readable_size(size, decimal_places=2):
+    for unit in ['B', 'KB', 'MB', 'GB', 'TB']:
+        if size < 1024.0:
+            break
+        size /= 1024.0
+    return f"{size:.{decimal_places}f}{unit}"
+
+
+def json_encode_provenance(provenance_dict, num_replicates=1):
+    """
+        Return a JSON representation of the provenance
+    """
+    prov = ProvenanceEncoderDecoder().encode(provenance_dict)
+    if len(prov) * num_replicates > 2_097_152:
```

Comment:
I don't think we need to worry about the number of replicates. If someone does 10**6 replicates, they're almost certainly not storing the actual files, so it won't matter (and this warning will worry users who haven't seen it before).

I think we just need to worry about the case in which a single rep goes over the threshold. (2M seems as good as any)

Reply:
I'm not so sure about this - even if they are not storing the files, it's still using up RAM. A simple, standard provenance is ~1500 chars, so the 10**6 replicate case would result in 1.5GB of provenance data in RAM without warning. 

I do get that anyone that uses a large number of replicates will always see this message though, which is not ideal. I propose we have 2M limit for single replicate and 500M total, and make the warning less scary by adding: "This is nothing to worry about unless you wish to save memory/storage"

Reply:
The standard approach for doing lots of replicates is 
```python
for ts in msprime.simulate(10, num_replicates=10**6):
    # do something with ts
```
So, only one tree sequence (and therefore provenance) is in memory at any one time. Unless you store the replicates in a list (which is a terrible idea anyway) the product of num_reps x provenance size isn't going to matter.

Unless I'm missing something here?

Reply:
picard_facepalm.gif 
Of course!